text/0000-async.md

```diff
@@ -0,0 +1,774 @@
+for helping detect bugs:
+
+```rust
+/// Marks the current thread as being within the dynamic extent of an
+/// executor.
+///
+/// Executor implementations should call this function before beginning to
+/// execute a tasks, and drop the returned `Enter` value after completing
+/// task execution:
+///
+/// ```rust
+/// let enter = enter().expect("...");
+/// /* run task */
+/// drop(enter);
+/// ```
+///
+/// Doing so ensures that executors aren't accidentally invoked in a nested fashion.
+/// When that happens, the inner executor can block waiting for an event that can
+/// only be triggered by the outer executor, leading to a deadlock.
+///
+/// # Error
+///
+/// Returns an error if the current thread is already marked, in which case the
+/// caller should panic with a tailored error message.
+pub fn enter() -> Result<Enter, EnterError>
+```
+
+As stated in the doc comment, the expectation is that all executors will wrap
+their task execution within an `enter` to detect inadvertent nesting.
+
+### Task contexts
+
+All tasks are executed with two pieces of contextual information:
+
+- A `Waker` for waking the task up later on.
+- An executor, which is the "default" place to spawn further tasks.
+
+Notably, this list does *not* include task-local data; that can be addressed
+externally, as we'll see in a later section.
+
+The `core::task::Context` type gathers the (stack-rooted) contextual information
+together, and is passed by mutable reference to all polling functions:
+
+```rust
+/// Information about the currently-running task.
+///
+/// Contexts are always tied to the stack, since they are set up specifically
+/// when performing a single `poll` step on a task.
+pub struct Context<'a> { .. }
+
+impl<'a> Context<'a> {
+    pub fn new(waker: &'a Waker, executor: &'a mut Executor) -> Context<'a>
+
+    /// Get the `Waker` associated with the current task.
+    pub fn waker(&self) -> &Waker;
+
+    /// Run an asynchronous computation to completion on the default executor.
+    ///
+    /// # Panics
+    ///
+    /// This method will panic if the default executor is unable to spawn.
+    /// To handle executor errors, use the `executor` method instead.
+    pub fn spawn(&mut self, f: impl Async<Output = ()> + 'static + Send);
+
+    /// Get the default executor associated with this task.
+    ///
+    /// This method is useful primarily if you want to explicitly handle
+    /// spawn failures.
+    ///
+    /// NB: this will remain unstable until the final `Executor` trait is ready.
+    pub fn executor(&mut self) -> &mut BoxExecutor;
+}
+```
+
+Note that the `spawn` method here will box until `Executor` is added.
+
+## `core::ops` module
+
+With all of the above task infrastructure in place, defining `Async` is
+straightforward. Since it's a "lang item", i.e. a trait that the language itself
+treats specially when compiling `async` blocks, the trait goes into the
+`core::ops` module, much as with the `Fn` family of traits.
```

Comment:
It would then be helpful to mention in the docs that one way to look at `Async` is it being the trait that overloads "operator `await`", much like `Fn` traits overload "operator `()`".

Reply:
I regret not calling `Fn` `Call`, which makes me wonder if `Async` should be `Await`.

Reply:
> But Fn* doesn't just represent that a type is Callable. Even though one could implement Fn* for a custom type (but not on today's Rust, IIRC), in the vast majority of cases that's not what happens. What happens is a type implementing Fn* is a function and its type carries no additional meaning beyond that (so Fn* is a type of trait you'd use with impl Trait) and has no API over that. And that is why, I believe, it makes sense to call it not just Callable, but a Function.

I think this is retconning the mistake that we haven't yet stabilized the mechanism behind the `Fn` trait being variadic and therefore they can't be implemented on stable.
It was *certainly* a regrettable mistake IMO (cc @nikomatsakis).